### PR TITLE
UML-1256 - moved the status check to api side and returns only LPA whose status meets

### PR DESCRIPTION
**Pull request by SeemaMenon**
# Purpose
https://opgtransform.atlassian.net/browse/UML-1256

Fixes UML-1256

## Approach

_Explain how your code addresses the purpose of the change_

## Learning

_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about the Use a Lasting Power of Attorney service_

## Checklist

* [X] I have performed a self-review of my own code
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [X] I have added tests to prove my work
* [ ] I have added welsh translation tags and updated translation files
* [ ] I have run an accessibility tool on any pages I have made changes to and fixed any issues found
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-api/app/src/App/src/Service/Lpa/IsValidLpa.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     *
+     *
+     * @param array $lpa An LPA data structure
+     * @param string $actorId The actors Database ID or Sirius UId to search for within the $lpa data structure
+     *
+     * @return ?bool True if status is Registered or Cancelled
+     */
+    public function __invoke(array $lpa): ?bool
+    {
+        if (
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/IsValidLpa.php:34]: this login can be simplified to `return strtolower($lpa['status']) === self::LPA_REGISTERED || strtolower($lpa['status']) === self::LPA_CANCELLED`

File: service-api/app/src/App/src/Service/Lpa/LpaService.php [hunk 2]
```diff
@@ -101,19 +104,25 @@ public function getByUserLpaActorToken(string $token, string $userId): ?array
...
+        if (($this->isValidLpa)($lpaData)) {
+            return [
+                'user-lpa-actor-token' => $map['Id'],
+                'date' => $lpa->getLookupTime()->format('c'),
+                'actor' => $actor,
+                'lpa' => $lpaData,
+            ];
+        }
+
+        return [];
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:125]: how come you return an empty array instead of null?

File: service-api/app/test/AppTest/Service/Lpa/IsValidLpaTest.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        $this->loggerProphecy = $this->prophesize(LoggerInterface::class);
+        $this->isValidLpaProphecy = $this->prophesize(IsValidLpa::class);
+    }
+
+    private function isValidLpaResolver(): IsValidLpa
+    {
+        return new IsValidLpa(
+            $this->loggerProphecy->reveal(),
+            $this->isValidLpaProphecy->reveal()
```
Comment (GemTay) [inline service-api/app/test/AppTest/Service/Lpa/IsValidLpaTest.php:33]: You only need to pass the logger into the constructor here

File: service-api/app/test/AppTest/Service/Lpa/IsValidLpaTest.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        $lpa = [
+            'status' => 'Registered',
+            'donor' => [
+                'id' => 1,
+            ]
+        ];
+
+        $resolver = $this->isValidLpaResolver();
+        $result = $resolver($lpa);
+        $this->assertEquals(true, $result);
```
Comment (GemTay) [inline service-api/app/test/AppTest/Service/Lpa/IsValidLpaTest.php:49]: using assertEquals is fine here, but you could use the assertTrue and assertFalse methods too :)

File: service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php [hunk 5]
```diff
@@ -235,6 +245,12 @@ private function init_valid_user_token_test(): stdClass
                 ]
             ]);
 
+        // check valid lpa
+        $this->isValidLpaProphecy->__invoke(
+            $t->Lpa->getData()
+        )->willReturn(true);
+        //
```
Comment (GemTay) [inline service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php:252]: empty comment to be removed

File: service-front/app/features/actor-view-lpa.feature [hunk 6]
```diff
@@ -15,9 +15,18 @@ Feature: View an LPA that I have added to my account
...
-      | Revoked   | This LPA has been cancelled |
+
+   @integration @ui
+   Scenario Outline: The user can view an LPA added to their account
+     Given I am on the dashboard page
+     When I request to view an LPA which status is "<status>"
+     Then I am taken back to the dashboard page
+    Examples:
+      | status  | message |
+      | Revoked |This LPA has been revoked |
```
Comment (GemTay) [inline service-front/app/features/actor-view-lpa.feature:30]: Since you're only passing 1 example, remove the examples and have the status in the step definition `When I request to view an LPA which status is revoked`

File: service-front/app/features/context/UI/LpaContext.php [hunk 7]
```diff
@@ -2415,6 +2434,15 @@ public function theFullLPAIsDisplayedWithTheCorrect($message)
         $this->ui->assertPageContainsText($message);
     }
 
+    /**
+     * @Then /^The (.*) LPA details and (.*) are not displayed$/
+     */
+    public function theRevokedLPADetailsAndMessageAreNotDisplayed($status, $message)
```
Comment (GemTay) [inline service-front/app/features/context/UI/LpaContext.php:2440]: i cant see this step in the feature file?

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php [hunk 8]
```diff
@@ -57,13 +59,22 @@ public function handle(ServerRequestInterface $request): ResponseInterface
         $user = $this->getUser($request);
         $identity = (!is_null($user)) ? $user->getIdentity() : null;
 
+        //UML-1394 TO BE REMOVED IN FUTURE TO SHOW PAGE NOT FOUND WITH APPROPRIATE CONTENT
         $lpaData = $this->lpaService->getLpaById($identity, $actorLpaToken);
+        if (count($lpaData) === 0) {
```
Comment (GemTay) [inline service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:64]: if you are returning an empty array in the api, this if statement should be `if (empty($lpaData))`

**Review by GemTay — CHANGES_REQUESTED** [review]
few small changes :)

**Review by SeemaMenon — COMMENTED** [review]
> few small changes :)

All done :)

File: service-api/app/features/actor-view-lpa.feature [hunk 9]
```diff
@@ -19,4 +19,12 @@ Feature: View an LPA that I have added to my account
...
-      | Revoked   | This LPA has been cancelled |
+
+  @integration @acceptance
+  Scenario Outline: The user can view an LPA added to their account
+    Given I am on the dashboard page
+    When I request to view an LPA which status is "<status>"
+    Then I am taken back to the dashboard page
+    Examples:
+      | status    |
+      | Revoked   |
```
Comment (GemTay) [inline service-api/app/features/actor-view-lpa.feature:30]: again, don't use examples where you are only providing 1

File: service-front/app/features/context/Integration/LpaContext.php [hunk 10]
```diff
@@ -1008,28 +1008,47 @@ public function iRequestToGiveAnOrganisationAccessToOneOfMyLPAs()
...
-                    StatusCodeInterface::STATUS_OK,
-                    [],
-                    json_encode(
-                        [
-                            'user-lpa-actor-token' => $this->actorLpaToken,
-                            'date' => 'date',
-                            'lpa' => $this->lpa,
-                            'actor' => $this->lpaData['actor'],
-                        ]
+        if ($status === "Revoked") {
```
Comment (GemTay) [inline service-front/app/features/context/Integration/LpaContext.php:1017]: you're no longer passing in revoked as an example in the feature file so this isnt needed

File: service-front/app/features/context/UI/LpaContext.php [hunk 11]
```diff
@@ -2007,31 +2007,49 @@ public function iRequestToGoBackAndTryAgain()
...
-                    StatusCodeInterface::STATUS_OK,
-                    [],
-                    json_encode(
-                        [
-                            'user-lpa-actor-token' => $this->userLpaActorToken,
-                            'date' => 'date',
-                            'lpa' => $this->lpa,
-                            'actor' => $this->lpaData['actor'],
-                        ]
+        if($status === 'Revoked'){
```
Comment (GemTay) [inline service-front/app/features/context/UI/LpaContext.php:2017]: same here

File: service-front/app/features/context/Integration/AccountContext.php [hunk 12]
```diff
@@ -1037,4 +1037,12 @@ protected function prepareContext(): void
         $this->lpaFactory = $this->container->get(LpaFactory::class);
         $this->viewerCodeService = $this->container->get(ViewerCodeService::class);
     }
+
+    /**
+     * @Then /^The (.*) LPA details and (.*) message are not displayed$/
+     */
+    public function theRevokedLPADetailsAndMessageAreNotDisplayed($status, $message)
```
Comment (GemTay) [inline service-front/app/features/context/Integration/AccountContext.php:1044]: this step definition doesnt match the feature file?

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 8 above)
Comment (GemTay) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:64]: this still needs changing @SeemaMenon

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php [hunk 13]
```diff
@@ -57,13 +59,22 @@ public function handle(ServerRequestInterface $request): ResponseInterface
...
+            return $this->redirectToRoute('lpa.dashboard');
+        }
```
Comment (GemTay) [inline service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:66]: tag this handler with @codeCoverageIgnore to get rid of these codecov warnings

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 8 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:64]: That gives me an error when tried with  empty function. Hence kept it as it is ! Would you mind having a look please

File: service-front/app/features/context/Integration/AccountContext.php (hunk 12 above)
Comment (SeemaMenon) [reply service-front/app/features/context/Integration/AccountContext.php:1044]: got you, corrected!

File: service-front/app/features/context/Integration/LpaContext.php (hunk 10 above)
Comment (SeemaMenon) [reply service-front/app/features/context/Integration/LpaContext.php:1017]: That's right. But I am using same step definition for the test step below [for the revoked case]. Attached snapshot below:
Screenshot 2021-04-20 at 14 24 39

File: service-front/app/features/context/UI/LpaContext.php (hunk 11 above)
Comment (SeemaMenon) [reply service-front/app/features/context/UI/LpaContext.php:2017]: same explanation as above, if that makes sense.

File: service-api/app/features/actor-view-lpa.feature (hunk 9 above)
Comment (SeemaMenon) [reply service-api/app/features/actor-view-lpa.feature:30]: Amended..

File: service-api/app/src/App/src/Service/Lpa/LpaService.php (hunk 2 above)
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:125]: Theres a codecov warning here, so it looks like you need to add a test to the LpaServiceTest for the case where IsLpaValid returns an empty array

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 13 above)
Comment (GemTay) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:68]: This redirect needs to be in place for the other pages too, eg code make and check access codes

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php [hunk 14]
```diff
@@ -9,6 +9,8 @@
 use Common\Handler\Traits\User;
 use Common\Handler\UserAware;
 use Common\Service\Lpa\LpaService;
+use Mezzio\Flash\FlashMessageMiddleware;
+use Mezzio\Flash\FlashMessagesInterface;
```
Comment (GemTay) [inline service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:13]: these imports arent needed

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 13 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:68]: Done now :)

File: service-api/app/src/App/src/Service/Lpa/LpaService.php (hunk 2 above)
Comment (SeemaMenon) [reply service-api/app/src/App/src/Service/Lpa/LpaService.php:125]: Ok, added in now.

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 14 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php:13]: my bad, clean up done...

File: service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php (hunk 8 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/src/Handler/ViewLpaSummaryHandler.php]: This is changed now

File: service-api/app/src/App/src/Service/Lpa/IsValidLpa.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+class IsValidLpa
+{
+    private const LPA_REGISTERED = 'registered';
+    private const LPA_CANCELLED = 'cancelled';
+
+    private LoggerInterface $logger;
+    public function __construct(
+        LoggerInterface $logger
+    ) {
+        $this->logger = $logger;
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/IsValidLpa.php:18]: this logger isnt currently being used, so maybe add in a log message for when the status is something other than registered or cancelled

File: service-front/app/features/context/UI/LpaContext.php [hunk 16]
```diff
@@ -2380,6 +2398,17 @@ public function theFullLPAIsDisplayedWithTheCorrect($message)
...
     }
 
+    /**
+     * @Then /^The Revoked LPA details are not displayed$/
+     */
+    public function theRevokedLPADetailsAreNotDisplayed()
+    {
+        $this->ui->assertPageAddress('/lpa/dashboard');
+        $this->ui->assertPageNotContainsText(
+            $this->lpa->donor->salutation . ' ' . $this->lpa->donor->firstname . ' ' . $this->lpa->donor->middlenames . ' ' . $this->lpa->donor->surname
```
Comment (GemTay) [inline service-front/app/features/context/UI/LpaContext.php:2408]: The donor salutation isnt displayed on the dashboard, so can be removed here

**Review by GemTay — CHANGES_REQUESTED** [review]
couple of things

File: service-api/app/src/App/src/Service/Lpa/IsValidLpa.php (hunk 15 above)
Comment (SeemaMenon) [reply service-api/app/src/App/src/Service/Lpa/IsValidLpa.php:18]: ok

File: service-api/app/src/App/src/Service/Lpa/IsValidLpa.php [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+     *
+     * @return ?bool True if status is Registered or Cancelled
+     */
+    public function __invoke(array $lpa): ?bool
+    {
+        if (!(strtolower($lpa['status']) === self::LPA_REGISTERED || strtolower($lpa['status']) === self::LPA_CANCELLED)) {
+
+            $this->logger->info(
+                'The status of the LPA is  - {status}',
+                ['status' => $lpa['status']]
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/IsValidLpa.php:38]: We need to have a bit more context around which LPA this is, so maybe something like "LPA with id {lpaUid} has an invalid status of {status}" - and change the log level to notice

Comment (SeemaMenon) [reply service-api/app/src/App/src/Service/Lpa/IsValidLpa.php]: OK, done.

**Review by GemTay — APPROVED** [review]
Nice!
